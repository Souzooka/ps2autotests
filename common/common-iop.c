#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
