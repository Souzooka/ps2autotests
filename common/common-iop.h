#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
