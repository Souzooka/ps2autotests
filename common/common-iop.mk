# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
