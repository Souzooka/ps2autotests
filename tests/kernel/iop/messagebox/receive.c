#include <common-iop.h>
#include <thbase.h>
#include <thmsgbx.h>
#include "mbx-common.h"

#define STACK_SIZE 0x800

typedef struct
{
	s32 id;
	s32 mbxId;
} THREAD_PARAM;

void messageWaitThreadProc(THREAD_PARAM *param) {
	MSG* msg = NULL;
	ReceiveMbx((void**)&msg, param->mbxId);
	schedf("(thread: %d -> message: %d) ", param->id, msg->payload);
}

s32 createTestThread(u32 priority) {
	iop_thread_t threadInfo = {};
	threadInfo.attr      = TH_C;
	threadInfo.thread    = (void*)&messageWaitThreadProc;
	threadInfo.priority  = priority;
	threadInfo.stacksize = STACK_SIZE;
	return CreateThread(&threadInfo);
}

void testMultipleReceive(u32 mbxAttr) {
	static const int msgCount = 3;
	int i = 0;
	MSG msg[msgCount];
	THREAD_PARAM threadParams[msgCount];
	s32 threadIds[msgCount];
	
	iop_mbx_t mbxInfo = {};
	mbxInfo.option = 0;
	mbxInfo.attr   = mbxAttr;
	s32 mbxId = CreateMbx(&mbxInfo);

	for(i = 0; i < msgCount; i++) {
		memset(&threadParams[i], 0, sizeof(MSG));
		threadParams[i].id = i + 1;
		threadParams[i].mbxId = mbxId;
		threadIds[i] = createTestThread(getThreadPriority(TH_SELF) - (i + 1));
		StartThread(threadIds[i], &threadParams[i]);
	}
	
	printf("  box contents: ");
	printMbx(mbxId);
	
	schedf("  order: ");
	for(i = 0; i < msgCount; i++) {
		memset(&msg[i], 0, sizeof(MSG));
		msg[i].header.priority = msgCount - i;
		msg[i].payload         = i + 1;
		SendMbx(mbxId, &msg[i]);
	}
	schedf("\n");
	flushschedf();
	
	for(i = 0; i < msgCount; i++) {
		TerminateThread(threadIds[i]);
		DeleteThread(threadIds[i]);
	}
	
	DeleteMbx(mbxId);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	printf("receiving (fifo thread queuing):\n");
	testMultipleReceive(MBA_THFIFO | MBA_MSFIFO);

	printf("receiving (priority thread queuing):\n");
	testMultipleReceive(MBA_THPRI | MBA_MSFIFO);
	
	printf("invalid params:\n");
	{
		int result = ReceiveMbx(NULL, ~0);
		printf("  receiving message from invalid box: %d\n", result);
	}
	
	printf("-- TEST END\n");
	return 0;
}
