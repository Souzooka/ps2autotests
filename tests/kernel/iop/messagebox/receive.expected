-- TEST BEGIN
receiving (fifo thread queuing):
  box contents: attr: 0, option: 0, wait threads: 3, messages: 0
  order: (thread: 1 -> message: 1) (thread: 2 -> message: 2) (thread: 3 -> message: 3) 
receiving (priority thread queuing):
  box contents: attr: 1, option: 0, wait threads: 3, messages: 0
  order: (thread: 3 -> message: 1) (thread: 2 -> message: 2) (thread: 1 -> message: 3) 
invalid params:
  receiving message from invalid box: -410
-- TEST END
